### PR TITLE
Auto-assessment

### PR DESCRIPTION
**Pull request by kjohn-msft**
This PR includes support for automatic assessment.

For all new sequence numbers, the entire environment state for auto-assessment is refreshed to the latest (all folder locations, sequence number, up-to-date code, etc.), and the service and associated timer is brought to the desired state in an idempotent fashion. 

The life cycle manager was extended to handle multiple core processes with the auto-assessment process when invoked/running always taking second place to mainline core invocation.

Generic service manager and timer manager classes were created that depend on a specific instantiation of the desired auto-assessment service.

### CODE REVIEW

File: src/extension/tests/TestExtConfigSettingsHandler.py [hunk 1]
```diff
@@ -359,6 +361,14 @@ def test_read_all_config_settings_from_file(self):
         self.assertNotEqual(config_settings.__getattribute__(self.config_public_settings_fields.maintenance_run_id), None)
         self.assertEqual(config_settings.__getattribute__(self.config_public_settings_fields.maintenance_run_id), "2019-07-22T12:12:14Z")
 
+        # verify assessmentMode is read successfully
+        self.assertNotEqual(config_settings.__getattribute__(self.config_public_settings_fields.assessment_mode), None)
+        self.assertEqual(config_settings.__getattribute__(self.config_public_settings_fields.assessment_mode), "AutomaticByPlatform")
+
+        # verify maximumAssessmentInterval is read successfully
+        self.assertNotEqual(config_settings.__getattribute__(self.config_public_settings_fields.maximum_assessment_interval), None)
+        self.assertEqual(config_settings.__getattribute__(self.config_public_settings_fields.maximum_assessment_interval), "3H")
```
Comment (rane-rajasi) [inline src/extension/tests/TestExtConfigSettingsHandler.py:370]: PT3H*

File: src/extension/tests/helpers/1234.settings [hunk 2]
```diff
@@ -14,7 +14,9 @@
                 "patchesToExclude": ["test", "*test"],
                 "internalSettings": "test",
                 "maintenanceRunId": "2019-07-20T12:12:14Z",
-                "patchMode": "AutomaticByPlatform"
+                "patchMode": "AutomaticByPlatform",
+                "assessmentMode": "AutomaticByPlatform",
+                "maximumAssessmentInterval": "3H"
```
Comment (rane-rajasi) [inline src/extension/tests/helpers/1234.settings:19]: PT3H*

**Review by rane-rajasi — COMMENTED** [review]
Review still in progress, submitting first iteration to unblock

File: src/core/src/core_logic/AutoAssessmentConfigurator.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        self.env_layer = env_layer
+        self.execution_config = execution_config
+
+        self.composite_logger = composite_logger
+        self.telemetry_writer = telemetry_writer
+        self.status_handler = status_handler
+
+
+    def create_configuration(self):
+        code, output = self.env_layer.run_command_output()
```
Comment (nikhim-um) [inline src/core/src/core_logic/AutoAssessmentConfigurator.py:30]: which command is being executed here?

Comment (nikhim-um) [reply src/core/src/core_logic/AutoAssessmentConfigurator.py:30]: I guess there should be at least one param here: run_command_output(self, cmd, no_output=False, chk_err=False)

Comment (kjohn-msft) [reply src/core/src/core_logic/AutoAssessmentConfigurator.py:30]: This entire class is dead code - it's not being used anywhere. It was part of a skeleton I initially wrote and got left behind. Thanks for catching it..

File: src/extension/tests/helpers/1234.settings (hunk 2 above)
Comment (kjohn-msft) [reply src/extension/tests/helpers/1234.settings]: Done

File: src/extension/src/ActionHandler.py [hunk 4]
```diff
@@ -223,6 +223,7 @@ def disable(self):
             self.setup(action=Constants.DISABLE, log_message="Disable triggered on extension")
             prev_patch_max_end_time = self.cmd_exec_start_time + datetime.timedelta(hours=0, minutes=Constants.DISABLE_MAX_RUNTIME)
             self.runtime_context_handler.process_previous_patch_operation(self.core_state_handler, self.process_handler, prev_patch_max_end_time, core_state_content=None)
+            # todo: disable auto-assessment here
```
[the remark this answers is not in the document]
Comment (kjohn-msft) [reply src/extension/src/ActionHandler.py]: Wasn't strictly necessary in the specific context of our extension's lifecycle and lifecycle management process. For defensive coding purposes, I've added in a faux disablement until control hands off fully to a new extension version.

File: src/extension/src/file_handlers/ExtConfigSettingsHandler.py [hunk 5]
```diff
@@ -130,11 +130,15 @@ def read_file(self, seq_no):
                     # todo: remove patch rollout id later
                     maintenance_run_id = self.get_ext_config_value_safely(config_settings_json, self.public_settings_all_keys.patch_rollout_id, raise_if_not_found=False)
                 patch_mode = self.get_ext_config_value_safely(config_settings_json, self.public_settings_all_keys.patch_mode, raise_if_not_found=False)
+                assessment_mode = self.get_ext_config_value_safely(config_settings_json, self.public_settings_all_keys.assessment_mode, raise_if_not_found=False)
+                maximum_assessment_interval = assessment_mode = self.get_ext_config_value_safely(config_settings_json, self.public_settings_all_keys.maximum_assessment_interval, raise_if_not_found=False)
```
[the remark this answers is not in the document]
Comment (kjohn-msft) [reply src/extension/src/file_handlers/ExtConfigSettingsHandler.py]: Addressed - good catch

File: src/core/src/service_interfaces/StatusHandler.py [hunk 6]
```diff
@@ -593,41 +615,54 @@ def add_error_to_status(self, message, error_code=Constants.PatchOperationErrorC
...
             if self.__installation_substatus_json is not None:
                 self.set_installation_substatus_json(status=self.__installation_substatus_json["status"], code=self.__installation_substatus_json["code"])
             else:
                 self.set_installation_substatus_json()
-        elif self.__current_operation == Constants.CONFIGURE_PATCHING:
-            self.__add_error(self.__configure_patching_errors, error_detail)
-            self.__configure_patching_total_error_count += 1
-            # retain previously set status, code and patchMode for configure patching substatus
+        elif current_operation == Constants.CONFIGURE_PATCHING or current_operation == Constants.CONFIGURE_PATCHING_AUTO_ASSESSMENT:
+            if current_operation == current_operation == Constants.CONFIGURE_PATCHING_AUTO_ASSESSMENT:
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:638]: current_operation == Constants.CONFIGURE_PATCHING_AUTO_ASSESSMENT:

**Review by rane-rajasi — COMMENTED** [review]
Review #2

File: src/core/src/core_logic/ServiceManager.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+    def stop_service(self):
+        code, out = self.__invoke_systemctl(self.service_stop_cmd.format(self.service_name), "Stopping the service.")
+        return code == 0
+
+    def reload_service(self):
+        code, out = self.__invoke_systemctl(self.service_reload_cmd.format(self.service_name), "Reloading the service.")
+        return code == 0
+
+    def get_service_status(self):
```
Comment (nikhim-um) [inline src/core/src/core_logic/ServiceManager.py:76]: Is command here is same as stop_service? both stop_service and get_service_status same command is used

File: src/extension/src/ProcessHandler.py [hunk 8]
```diff
@@ -87,6 +94,30 @@ def start_daemon(self, seq_no, config_settings, ext_env_handler):
...
+        self.logger.log_debug("Staging auto assessment shell script with latest config.")
+        try:
+            exec_dir = os.path.dirname(os.path.realpath(__file__))
+            auto_assess_sh_path = os.path.join(exec_dir, Constants.CORE_AUTO_ASSESS_SH_FILE_NAME)
+            self.logger.log_debug("Discovered auto_assess_sh_path. [Path={0}]".format(auto_assess_sh_path))
+
+            if os.path.exists(auto_assess_sh_path):
+                os.remove(auto_assess_sh_path)
+
+            core_process_command = str.replace(core_process_command, Constants.CORE_CODE_FILE_NAME, os.path.join(exec_dir, Constants.CORE_CODE_FILE_NAME))
```
Comment (nikhim-um) [inline src/extension/src/ProcessHandler.py:108]: Getting below exception while testing: 
Python version: 3.8.10 (default, Jun  2 2021, 10:49:15)  [GCC 9.4.0]
DEBUG: Staging auto assessment shell script with latest config.
DEBUG: Discovered auto_assess_sh_path. [Path=/var/lib/waagent/Microsoft.CPlat.Core.LinuxPatchExtension-1.6.19/MsftLinuxPatchAutoAssess.sh]
ERROR: Unable to stage auto-assess shim. [Error=descriptor 'replace' for 'str' objects doesn't apply to a 'list' object]

File: src/extension/src/Constants.py [hunk 9]
```diff
@@ -98,6 +99,7 @@ class TelemetryEventLevel(EnumBackport):
     PATCH_INSTALLATION_SUMMARY = "PatchInstallationSummary"
     CONFIGURE_PATCHING = "ConfigurePatching"
     CONFIGURE_PATCHING_SUMMARY = "ConfigurePatchingSummary"
+    AUTO_ASSESSMENT_ONLY = "-AutoAssessmentOnly"
```
[the remark this answers is not in the document]
Comment (kjohn-msft) [reply src/extension/src/Constants.py:102]: Dead code -- removed..

File: src/core/src/core_logic/ConfigurePatchingProcessor.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+            # disable auto OS updates if VM is configured for platform updates only.
+            # NOTE: this condition will be false for Assessment operations, since patchMode is not sent in the API request
+            if self.current_auto_os_patch_state == Constants.AutomaticOsPatchStates.ENABLED and self.execution_config.patch_mode == Constants.PatchModes.AUTOMATIC_BY_PLATFORM:
+                self.package_manager.disable_auto_os_update()
+
+            self.current_auto_os_patch_state = self.package_manager.get_current_auto_os_patch_state()
+        except Exception as error:
+            self.__report_consolidated_configure_patch_status(status=Constants.STATUS_ERROR, error=error)
+            self.configure_patching_successful &= False
```
[the remark this answers is not in the document]
Comment (kjohn-msft) [reply src/core/src/core_logic/ConfigurePatchingProcessor.py:76]: It's just as a developer hint that there are multiple values contributing to the outcome, but just evaluated at different points in the flow.

File: src/core/src/core_logic/ConfigurePatchingProcessor.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+            elif self.execution_config.assessment_mode == Constants.AssessmentModes.IMAGE_DEFAULT:
+                self.composite_logger.log_debug("Disabling platform-based automatic assessment.")
+                self.auto_assess_timer_manager.remove_timer()
+                self.auto_assess_service_manager.remove_service()
+                self.current_auto_assessment_state = Constants.AutoAssessmentStates.DISABLED
+            else:
+                raise Exception("Unknown assessment mode specified. [AssessmentMode={0}]".format())
+
+            self.__report_consolidated_configure_patch_status(status=Constants.STATUS_SUCCESS)
+            self.composite_logger.log_debug("Completed configuring automatic assessment mode.")
```
[the remark this answers is not in the document]
Comment (kjohn-msft) [reply src/core/src/core_logic/ConfigurePatchingProcessor.py:99]: Changed from configuring to processing configuration.

File: src/core/src/core_logic/TimerManager.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        self.telemetry_writer = telemetry_writer
+
+        self.service_name = service_info.service_name
+        self.service_desc = service_info.service_desc
+        self.service_exec_path = service_info.service_exec_path
+
+        self.__systemd_path = "/run/systemd/system/"
+        self.__systemd_timer_unit_path = "/etc/systemd/system/{0}.timer"
+
+        self.timer_start_cmd = "sudo systemctl start {0}.timer"
```
[the remark this answers is not in the document]
Comment (kjohn-msft) [reply src/core/src/core_logic/TimerManager.py:36]: Needs some more thought -- pushing to a future review as it's non-urgent.

File: src/extension/tests/TestExtConfigSettingsHandler.py [hunk 13]
```diff
@@ -352,6 +354,14 @@ def test_read_all_config_settings_from_file(self):
         self.assertNotEqual(config_settings.__getattribute__(self.config_public_settings_fields.maintenance_run_id), None)
         self.assertEqual(config_settings.__getattribute__(self.config_public_settings_fields.maintenance_run_id), "2019-07-20T12:12:14Z")
 
+        # verify assessmentMode is read successfully
+        self.assertNotEqual(config_settings.__getattribute__(self.config_public_settings_fields.assessment_mode), None)
+        self.assertEqual(config_settings.__getattribute__(self.config_public_settings_fields.assessment_mode), "AutomaticByPlatform")
```
[the remark this answers is not in the document]
Comment (kjohn-msft) [reply src/extension/tests/TestExtConfigSettingsHandler.py:359]: It failed when the prod code reverted so it's covered now..

File: src/core/src/core_logic/TimerManager.py [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+        self.env_layer = env_layer
+        self.execution_config = execution_config
+        self.composite_logger = composite_logger
+        self.telemetry_writer = telemetry_writer
+
+        self.service_name = service_info.service_name
+        self.service_desc = service_info.service_desc
+        self.service_exec_path = service_info.service_exec_path
+
+        self.__systemd_path = "/run/systemd/system/"
```
[the remark this answers is not in the document]
Comment (kjohn-msft) [reply src/core/src/core_logic/TimerManager.py:33]: Pulling out of constants now

File: src/core/src/core_logic/TimerManager.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+        self.timer_stop_cmd = "sudo systemctl stop {0}.timer"
+        self.timer_reload_cmd = "sudo systemctl reload-or-restart {0}.timer"
+        self.timer_enable_cmd = "sudo systemctl enable {0}.timer"
+        self.timer_disable_cmd = "sudo systemctl disable {0}.timer"
+        self.timer_is_enabled_cmd = "sudo systemctl is-enabled {0}.timer"
+        self.timer_is_active_cmd = "sudo systemctl is-active {0}.timer"
+        self.systemctl_daemon_reload_cmd = "sudo systemctl daemon-reload"
+
+    def check_timer_health(self):
+        pass
```
[the remark this answers is not in the document]
Comment (kjohn-msft) [reply src/core/src/core_logic/TimerManager.py:46]: This is a placeholder for later. Will cover in a second PR.

File: README.md [hunk 16]
```diff
@@ -56,11 +63,11 @@ and add this file into the configFolder path from HandlerEnvironment.json
 
 ## 3. Troubleshooting
 
-Within Azure VM, you can find logs/config files at these locations:
+Within your Azure VM, you can find logs/config files at these locations:
 
 * Agent log: `/var/log/waagent.log`
-* Extension logs: `under /var/log/azure/Microsoft.CPlat.Core.Edp.LinuxPatchExtension/`
-* Other Extension files (such as status blob, config file, etc): `/var/lib/waagent/Microsoft.CPlat.Core.Edp.LinuxPatchExtension-<version>/`
+* Extension logs: `under /var/log/azure/Microsoft.CPlat.Core.LinuxPatchExtension/`
```
Comment (rane-rajasi) [inline README.md:69]: nit: remove 'under'

File: src/core/src/core_logic/TimerManager.py [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+    def stop_timer(self):
+        code, out = self.__invoke_systemctl(self.timer_stop_cmd.format(self.service_name), "Stopping the timer.")
+        return code == 0
+
+    def reload_timer(self):
+        code, out = self.__invoke_systemctl(self.timer_reload_cmd.format(self.service_name),"Reloading the timer.")
+        return code == 0
+
+    def get_timer_status(self):
+        # code, out = self.__invoke_systemctl(self.timer_status_cmd.format(self.service_name), "Stopping the timer.")
```
Comment (rane-rajasi) [inline src/core/src/core_logic/TimerManager.py:105]: Highlighting this one to ensure it isn't missed in the next iteration

**Review by rane-rajasi — APPROVED** [review]
Approved with minor comments

File: src/core/src/core_logic/ServiceManager.py (hunk 7 above)
Comment (kjohn-msft) [reply src/core/src/core_logic/ServiceManager.py:76]: Adding later as not needed for mainline scenario..

**Pull request merged**